File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.py
from heapq import heappush, heapreplace


class KthLargest:
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.heap = []
        for num in nums:
            self.add(num)

    def add(self, val: int) -> int:
        if len(self.heap) < self.k:
            heappush(self.heap, val)
        elif val > self.heap[0]:
            heapreplace(self.heap, val)
        return self.heap[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.cpp
class KthLargest {
public:
    KthLargest(int k, vector<int>& nums) : k_(k) {
        for (int num : nums) {
            add(num);
        }
    }

    int add(int val) {
        if (heap_.size() < k_) {
            heap_.push(val);
        } else if (val > heap_.top()) {
            heap_.push(val);
            heap_.pop();
        }
        return heap_.top();
    }

private:
    int k_;
    priority_queue<int, vector<int>, greater<int>> heap_;
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */
